File: webapp/lib/replayer.ts
/**
 * TypeScript port of record_replay/replayer.py.
 *
 * Works on a COO (coordinate) representation in-memory.  Blob downloads are
 * done via the azure-blob helpers; dense conversion is deferred to the API
 * route so this layer stays pure data.
 *
 * Step semantics (identical to Python):
 *   step 0   = initial matrix
 *   step N   = state after N events applied
 *   eventAt(N) = the event that caused transition N-1 → N
 */

import { downloadBlob, downloadBlobOptional, listBlobs } from "./azure-blob";
import { decodeCsr, decodeEvents, CsrMatrix, EventRecord } from "./msgpack-decode";

// ---------------------------------------------------------------------------
// Working matrix representation — dense rows of { col → value } maps
// Mirrors the Python lil_matrix used during event replay.
// ---------------------------------------------------------------------------

type LilMatrix = Map<number, number>[];   // rows[r].get(col) → value

function csrToLil(csr: CsrMatrix): LilMatrix {
  const [nRows] = csr.shape;
  const rows: LilMatrix = Array.from({ length: nRows }, () => new Map());
  for (let r = 0; r < nRows; r++) {
    const start = csr.indptr[r];
    const end   = csr.indptr[r + 1];
    for (let k = start; k < end; k++) {
      rows[r].set(csr.indices[k], csr.data[k]);
    }
  }
  return rows;
}

function lilToCsr(rows: LilMatrix, shape: [number, number]): CsrMatrix {
  const dataArr: number[]  = [];
  const idxArr:  number[]  = [];
  const ptrArr:  number[]  = [0];

  for (const row of rows) {
    const sorted = [...row.entries()].sort(([a], [b]) => a - b);
    for (const [col, val] of sorted) {
      idxArr.push(col);
      dataArr.push(val);
    }
    ptrArr.push(dataArr.length);
  }

  return {
    shape,
    data:    new Float64Array(dataArr),
    indices: new Int32Array(idxArr),
    indptr:  new Int32Array(ptrArr),
  };
}

// ---------------------------------------------------------------------------
// Row operation engine (mirrors _apply_event in replayer.py)
// ---------------------------------------------------------------------------

function applyEvent(rows: LilMatrix, ev: EventRecord): void {
  const p = ev.params;

  switch (ev.event_type) {
    case "swap_rows": {
      const tmp = rows[p.row_i];
      rows[p.row_i] = rows[p.row_j];
      rows[p.row_j] = tmp;
      break;
    }
    case "scale_row": {
      const row = rows[p.row];
      for (const [col, val] of row) row.set(col, val * p.scalar);
      break;
    }
    case "add_scaled_row": {
      const tgt = rows[p.target];
      const src = rows[p.source];
      for (const [col, val] of src) {
        const cur = tgt.get(col) ?? 0;
        const next = cur + p.scalar * val;
        if (Math.abs(next) > 1e-14) tgt.set(col, next);
        else tgt.delete(col);
      }
      break;
    }
    case "xor_rows": {
      // F₂ XOR: symmetric difference of column sets — cols in both cancel to 0
      const tgt = rows[p.target];
      for (const [col] of rows[p.source]) {
        if (tgt.has(col)) tgt.delete(col);
        else tgt.set(col, 1);
      }
      break;
    }
  }
}

// ---------------------------------------------------------------------------
// COO export (what the API routes return to the frontend)
// ---------------------------------------------------------------------------

export interface CooMatrix {
  shape: [number, number];
  row:   number[];
  col:   number[];
  data:  number[];
}

function csrToCoo(csr: CsrMatrix): CooMatrix {
  const row: number[] = [];
  const col: number[] = [];
  const data: number[] = [];
  const [nRows] = csr.shape;

  for (let r = 0; r < nRows; r++) {
    const start = csr.indptr[r];
    const end   = csr.indptr[r + 1];
    for (let k = start; k < end; k++) {
      row.push(r);
      col.push(csr.indices[k]);
      data.push(csr.data[k]);
    }
  }
  return { shape: csr.shape, row, col, data };
}

function diffCells(a: CsrMatrix, b: CsrMatrix): [number, number][] {
  const cooA = csrToCoo(a);
  const cooB = csrToCoo(b);

  const mapA = new Map<string, number>();
  for (let i = 0; i < cooA.row.length; i++) {
    mapA.set(`${cooA.row[i]},${cooA.col[i]}`, cooA.data[i]);
  }

  const changed = new Set<string>();
  for (let i = 0; i < cooB.row.length; i++) {
    const key = `${cooB.row[i]},${cooB.col[i]}`;
    if (mapA.get(key) !== cooB.data[i]) changed.add(key);
    mapA.delete(key);
  }
  // Any key still in mapA had a nonzero in A but zero in B → changed
  for (const key of mapA.keys()) changed.add(key);

  return [...changed].map((k) => k.split(",").map(Number) as [number, number]);
}

// ---------------------------------------------------------------------------
// Session state — one instance per session, cached at module level
// ---------------------------------------------------------------------------

interface SessionState {
  events:      EventRecord[];
  initial:     CsrMatrix;
  checkpoints: Map<number, CsrMatrix>;  // step → matrix (lazy-loaded)
  ckptSteps:   number[];                // sorted checkpoint step numbers on disk
  shape:       [number, number];
}

// Module-level cache: survives across requests in `next start` Node.js mode.
const SESSION_CACHE = new Map<string, SessionState>();

async function loadSession(sessionId: string): Promise<SessionState> {
  if (SESSION_CACHE.has(sessionId)) return SESSION_CACHE.get(sessionId)!;

  const prefix = `sessions/${sessionId}/`;
  const blobs  = await listBlobs(prefix);

  const ckptSteps: number[] = [];
  for (const name of blobs) {
    const base = name.slice(prefix.length);
    const m = base.match(/^checkpoint_(\d+)\.msgpack$/);
    if (m) ckptSteps.push(parseInt(m[1], 10));
  }
  ckptSteps.sort((a, b) => a - b);

  const [initialBuf, commandsBuf] = await Promise.all([
    downloadBlob(`${prefix}initial.msgpack`),
    downloadBlob(`${prefix}commands.msgpack`),
  ]);

  const initial = decodeCsr(initialBuf);
  const events  = decodeEvents(commandsBuf);

  const state: SessionState = {
    events,
    initial,
    checkpoints: new Map(),
    ckptSteps,
    shape: initial.shape,
  };
  SESSION_CACHE.set(sessionId, state);
  return state;
}

async function getCheckpoint(state: SessionState, sessionId: string, step: number): Promise<CsrMatrix> {
  if (step === 0) return state.initial;
  if (state.checkpoints.has(step)) return state.checkpoints.get(step)!;

  const buf = await downloadBlob(`sessions/${sessionId}/checkpoint_${step}.msgpack`);
  const mat = decodeCsr(buf);
  state.checkpoints.set(step, mat);
  return mat;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export interface StepResult {
  step:         number;
  totalSteps:   number;
  matrix:       CooMatrix;
  changedCells: [number, number][];
  event:        EventRecord | null;
}

export async function getStep(sessionId: string, n: number): Promise<StepResult> {
  const state = await loadSession(sessionId);
  const total = state.events.length;
  n = Math.max(0, Math.min(n, total));

  // Find the largest checkpoint step ≤ n
  let bestCkpt = 0;
  for (const s of state.ckptSteps) {
    if (s <= n) bestCkpt = s;
    else break;
  }

  const base    = await getCheckpoint(state, sessionId, bestCkpt);
  const rows    = csrToLil(base);
  for (let i = bestCkpt; i < n; i++) applyEvent(rows, state.events[i]);
  const curr = lilToCsr(rows, state.shape);

  // Diff against previous step
  let prev: CsrMatrix;
  if (n === 0) {
    prev = curr;
  } else {
    const prevRows = csrToLil(curr);
    // undo last event to get prev state
    const prevBase = await getCheckpoint(state, sessionId, bestCkpt);
    const prevRowsFresh = csrToLil(prevBase);
    for (let i = bestCkpt; i < n - 1; i++) applyEvent(prevRowsFresh, state.events[i]);
    prev = lilToCsr(prevRowsFresh, state.shape);
  }

  return {
    step:         n,
    totalSteps:   total,
    matrix:       csrToCoo(curr),
    changedCells: n === 0 ? [] : diffCells(prev, curr),
    event:        n > 0 ? state.events[n - 1] : null,
  };
}

export interface Annotation {
  code_family?:  string;
  erasure_rate?: number;
  reorder?:      string;
  note?:         string;
}

export async function getSessionInfo(sessionId: string) {
  const [state, metaBuf] = await Promise.all([
    loadSession(sessionId),
    downloadBlobOptional(`sessions/${sessionId}/metadata.json`),
  ]);

  const annotation: Annotation | null = metaBuf
    ? (JSON.parse(metaBuf.toString("utf-8")) as Annotation)
    : null;

  return {
    sessionId,
    totalSteps:   state.events.length,
    nCheckpoints: state.ckptSteps.length,
    shape:        state.shape,
    annotation,
  };
}

File: record_replay/replayer.py
"""Replays a recorded Gaussian-elimination session one step at a time.

Public API
----------
replayer = Replayer(data_dir)
replayer.load_session(session_id)   → None
replayer.get_step(n)                → scipy.sparse.csr_matrix
replayer.total_steps()              → int
replayer.event_at(n)                → dict | None

Step semantics
--------------
Step 0  = initial matrix (no operations applied, from initial.msgpack)
Step N  = state after N operations have been applied
event_at(N) returns the event that caused the N-1 → N transition.

Checkpoint strategy
-------------------
get_step() finds the largest checkpoint C ≤ N, loads it as a baseline, then
replays events C..N-1 forward.  If the internal cache is already at a step M
with C ≤ M ≤ N, it uses M as the baseline instead, avoiding re-loading from
disk.  Sequential playback therefore costs exactly one event per step.

The working matrix during replay is kept as scipy.sparse.lil_matrix, which
supports efficient in-place row mutations.  The caller receives a CSR matrix;
dense conversion happens only in the viewer layer.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Optional

import scipy.sparse as sp

from storage import read_msgpack


class Replayer:
    def __init__(self, data_dir: str | Path):
        self._data_dir = Path(data_dir)
        self._session_dir: Optional[Path] = None
        self._initial: Optional[sp.csr_matrix] = None
        self._events: list[dict] = []
        self._ckpt_steps: list[int] = []   # sorted checkpoint step numbers on disk
        self._cache_step: Optional[int] = None
        self._cache_mat: Optional[sp.csr_matrix] = None

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def load_session(self, session_id: str) -> None:
        """Load all session data from disk; resets the internal cache."""
        self._session_dir = self._data_dir / f"session_{session_id}"
        self._initial = read_msgpack(self._session_dir / "initial.msgpack")
        self._events = read_msgpack(self._session_dir / "commands.msgpack")

        self._ckpt_steps = sorted(
            int(m.group(1))
            for p in self._session_dir.iterdir()
            if (m := re.fullmatch(r"checkpoint_(\d+)\.msgpack", p.name))
        )

        # Prime the cache at step 0 so the first get_step(0) is a hit
        self._cache_step = 0
        self._cache_mat = self._initial.copy()

    def total_steps(self) -> int:
        """Number of recorded events (slider runs 0 … total_steps)."""
        return len(self._events)

    def get_step(self, n: int) -> sp.csr_matrix:
        """Return the sparse CSR matrix after n operations have been applied.

        Never converts to dense; that is deferred to the rendering layer.
        """
        n = max(0, min(n, self.total_steps()))

        if self._cache_step == n:
            return self._cache_mat

        # Largest checkpoint step ≤ n  (step 0 == initial.msgpack)
        best_ckpt = 0
        for s in self._ckpt_steps:
            if s <= n:
                best_ckpt = s
            else:
                break

        # Prefer the cache over re-loading from disk when it is a valid start
        if (self._cache_step is not None
                and best_ckpt <= self._cache_step <= n):
            mat = self._cache_mat.tolil()
            start = self._cache_step
        else:
            mat = self._load_checkpoint(best_ckpt).tolil()
            start = best_ckpt

        for i in range(start, n):
            _apply_event(mat, self._events[i])

        result = mat.tocsr()
        self._cache_step = n
        self._cache_mat = result
        return result

    def event_at(self, n: int) -> Optional[dict]:
        """Event that caused the (n-1) → n transition, or None for step 0."""
        if n <= 0 or n > len(self._events):
            return None
        return self._events[n - 1]

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _load_checkpoint(self, step: int) -> sp.csr_matrix:
        if step == 0:
            return self._initial.copy()
        return read_msgpack(self._session_dir / f"checkpoint_{step}.msgpack")


# ---------------------------------------------------------------------------
# Row-operation engine — works on lil_matrix in-place for full sparsity
# ---------------------------------------------------------------------------

def _apply_event(mat: sp.lil_matrix, ed: dict) -> None:
    etype = ed["event_type"]
    p = ed["params"]

    if etype == "swap_rows":
        i, j = p["row_i"], p["row_j"]
        mat.rows[i], mat.rows[j] = mat.rows[j], mat.rows[i]
        mat.data[i], mat.data[j] = mat.data[j], mat.data[i]

    elif etype == "scale_row":
        row, scalar = p["row"], p["scalar"]
        mat.data[row] = [v * scalar for v in mat.data[row]]

    elif etype == "add_scaled_row":
        target, source, scalar = p["target"], p["source"], p["scalar"]
        # Merge scalar * source into target using a column-keyed dict so that
        # column index ordering and fill-in are handled correctly.
        tgt = dict(zip(mat.rows[target], mat.data[target]))
        for col, val in zip(mat.rows[source], mat.data[source]):
            tgt[col] = tgt.get(col, 0.0) + scalar * val
        # Drop numerical zeros (same threshold as the GE generator)
        tgt = {c: v for c, v in tgt.items() if abs(v) > 1e-14}
        if tgt:
            pairs = sorted(tgt.items())
            mat.rows[target] = [c for c, _ in pairs]
            mat.data[target] = [v for _, v in pairs]
        else:
            mat.rows[target] = []
            mat.data[target] = []

    elif etype == "xor_rows":
        target, source = p["target"], p["source"]
        # Symmetric difference of column sets: cols in exactly one of the two
        # rows become 1; cols in both cancel to 0 (F₂ XOR semantics).
        tgt_cols = set(mat.rows[target])
        src_cols = set(mat.rows[source])
        result = sorted(tgt_cols.symmetric_difference(src_cols))
        mat.rows[target] = result
        mat.data[target] = [1.0] * len(result)
